src/python/WMCore/MicroService/Unified/MSMonitor.py

```diff
@@ -0,0 +1,196 @@
+        wname = reqName
+        # get existing transfer IDs record
+        doc = self.getTransferInfo(wname)
+        # loop over all records and obtain new transfer IDs
+        for rec in doc['transfers']:
+            dataset = rec['dataset']
+            dtype = rec['dataType']
+            # obtain new transfer ids for given request name and dataset
+            tids = self.getTransferIds(dataset)
+            rec = {'dataset': dataset, 'dataType': dtype, 'transferIDs': tids}
+            statusRecords += rec
+        return statusRecords
+
+    def pileupTransfersCompleted(self, reqName):
+        "Check if pileup transfers are completed for given request"
+        # in our case workflow name is the same as request name
+        wname = reqName
+        doc = self.getTransferInfo(wname)
+        # loop over all records and obtain new transfer IDs
+        for rec in doc['transfers']:
+            dataset = rec['dataset']
+            ctype = rec['dataType']
+            if ctype == 'pileup':
+                tids = self.getTransferIds(dataset)
+                # if we have any transfer IDs then we still have transfers
+                if tids:
+                    return False
+        return True
+
+    def checkStatusRecords(self, records):
+        """
+        Helper function to check status records (transfers)
+        :param records: list of status records
+        :return: new set and comlpeted status
+        """
+        completed = 0
+        new_records = []
+        for rec in records:
+            dataset = rec['dataset']
+            old_tids = rec['transferIDs']
+            new_tids = self.getTransferIds(dataset)
+            rec['transferIDs'] = new_tids
+            new_records.append(rec)
+            completed = 100 * float(len(new_tids)) / float(len(old_tids))
+        return new_records, completed
+
+    def updateTransferInfo(self, requestStatuses):
+        "Update transfer ids in backend"
+        tstamp = time.time()
+        for wname, statusRecord in requestStatuses:
+            doc = self.getTransferInfo(wname)
+            records = doc.get('transfers', [])
+            records += statusRecord
+            doc['workflowName'] = wname
+            doc['timestamp'] = tstamp
+            doc['transfers'] = records
+            self.reqmgrAux.postTransferInfo(wname, doc)
+
+    def getTransferInfo(self, wname):
+        """
+        Get transfer document from backend. The document has the following form:
+
+        .. doctest::
+
+            {"workflowName": "bla",
+             "timestamp": 123,
+             "transfers": [rec1, rec2, ... ]
+            }
+            where each record has the following format:
+            {"dataset":"/a/b/c", "dataType": "primary", "transferIDs": [1,2]}
+
+
+        :param wname: workflow name
+        :return: a tranfer ids document
+        """
+        return self.reqmgrAux.getTransferInfo(wname)
+
+    def getTransferIds(self, dataset):
```

Comment:
This should be called `getTransferStatus` or something like that. We'll provide a transfer/request ID, and get the transfer status back (method input is wrong as well).

Reply:
the function literally returns transfer Ids, why do you need to rename it? I don't mind but I want to understand your logic.

And, if we'll provide request ID, then I need to know two things:
- how to get requestID  from WMA request name
- and, I think we should store this requestID in transfer records, otherwise we lost this info and can't use it. So far we store dataset, timstamp, transferIDs

Once you'll provide this info I'll be happy to change the name of API and its input/logic to use requestID

Reply:
This is exactly what we discussed over that long private thread.
Things from the top of my head - but better getting back there to review what I say - the PhEDEx `subscribe` API returns the request ID (which I also call transfer ID).

As we discussed about the transfer document schema, this request ID is supposed to be persisted in Couch, according to this structure:
https://gist.github.com/amaltaro/72599f995b37a6e33566f3c749143154#file-transferor_format-json-L32

Reply:
Not so fast. Did you ever check what phedex subscriptions API return for given request? Please give it a try, e.g. this url (using dataset input parameter):
https://cmsweb-testbed.cern.ch/phedex/datasvc/xml/prod/subscriptions?dataset=/EG/Run2010A-WZEG-Nov4Skim_v1/RAW-RECO
will return proper subscription detials in no time (call time is 0.016 sec), while this call (using request ID)
https://cmsweb-testbed.cern.ch/phedex/datasvc/xml/prod/subscriptions?request=191805
will return no results in 45 seconds. May be I'm doing something wrong but so far usage of request ID does not seem to work with Phedex subscriptions API

Reply:
Interesting, I thought we could simply provide the request ID parameter to that API, but indeed it takes a long time to resolve it (even with a newer example/subscription).
It looks like we need to provide either the dataset or the block name, together with the request ID. In that case, we can quickly get a "valid" output, e.g.:
https://cmsweb-testbed.cern.ch/phedex/datasvc/json/prod/subscriptions?request=1857599&block=/RelValH125GGgluonfusion_14/Integ_Test-SC_LumiMask_PhEDEx_HG1908_Val_Alanv11-v11/NANOAODSIM%23*

We might have to ping Natalia on this item once again.

Reply:
so, what should we do? So far I left everything as is, i.e. I used dataset as input parameter. Until you clarify how this should be done I don't think we need to change anything.